core/modules/wdb/ChunkResource.cc

```diff
@@ -307,96 +311,81 @@ class ChunkEntry {
-
-    /// precondition: _mapMutex is held (locked by the caller)
-    /// Get the ChunkEntry map for a db, creating if necessary
-    Map& _getMap(std::string const& db) {
-        DbMap::iterator it = _dbMap.find(db);
-        if (it == _dbMap.end()) {
-            DbMap::value_type v(db, Map());
-            _dbMap.insert(v);
-            it = _dbMap.find(db);
-        }
-        return it->second;
-    }
-    /// precondition: _mapMutex is held (locked by the caller)
-    /// Get the ChunkEntry for a chunkId, creating if necessary
-    ChunkEntry& _getChunkEntry(Map& m, int chunkId) {
-        Map::iterator it = m.find(chunkId); // Select chunkId
-        if (it == m.end()) { // Insert if not exist
-            Map::value_type v(
-                              chunkId,
-                              std::make_shared<ChunkEntry>(chunkId)
-                             );
-            m.insert(v);
-            return *(v.second.get());
-        }
-        return *(it->second.get());
+
+ChunkResource ChunkResourceMgr::acquire(std::string const& db, int chunkId,
+                                        DbTableSet const& dbTableSet, IntVector const& subChunks) {
+    ChunkResource cr(this, new ChunkResource::Info(db, chunkId, dbTableSet, subChunks));
+    return cr;
+}
+
+
+void ChunkResourceMgr::release(ChunkResource::Info const& i) {
+     std::lock_guard<std::mutex> lock(_mapMutex);
+     Map& map = _getMap(i.db);
+     ChunkEntry& ce = _getChunkEntry(map, i.chunkId);
+     ce.release(i.db, i.tables, i.subChunkIds, _backend);
+}
+
+
+void ChunkResourceMgr::acquireUnit(ChunkResource::Info const& i) {
+    std::lock_guard<std::mutex> lock(_mapMutex);
+    Map& map = _getMap(i.db); // Select db
+    ChunkEntry& ce = _getChunkEntry(map, i.chunkId);
+    // Actually acquire
+    LOGS(_log, LOG_LVL_DEBUG, "acquireUnit info=" << i);
+    ce.acquire(i.db, i.tables, i.subChunkIds, _backend);
+}
+
+
+int ChunkResourceMgr::getRefCount(std::string const& db, int chunkId) {
+    std::lock_guard<std::mutex> lock(_mapMutex);
+    Map& map = _getMap(db); // Select db
+    Map::iterator it = map.find(chunkId); // Select chunkId
+    if (it == map.end()) {
+        return 0;
     }
+    ChunkEntry& ce =  *(it->second.get());
+    return ce.getRefCount();
+}
 
-    friend class ChunkResourceMgr;
-    DbMap _dbMap;
-    // Consider having separate mutexes for each db's map if contention becomes
-    // a problem.
-    std::shared_ptr<SQLBackend> _backend;
-    std::mutex _mapMutex; // Do not alter map without this mutex
-};
 
-////////////////////////////////////////////////////////////////////////
-// ChunkResourceMgr
-////////////////////////////////////////////////////////////////////////
+ChunkResourceMgr::Map& ChunkResourceMgr::_getMap(std::string const& db) {
+    DbMap::iterator it = _dbMap.find(db);
+    if (it == _dbMap.end()) {
+        DbMap::value_type v(db, Map());
+        _dbMap.insert(v);
+        it = _dbMap.find(db);
```

Comment:
`insert()` returns iterator (and bool) so you do not need`find()`. 
An the whole method is more or less equivalent to `return _dbMap[db];`